File: 0994-rotting-oranges/0994-rotting-oranges.cpp
class Solution{
public:
    // time complexity -> O(N*M + 4(N*M))
    // space complexity -> O(N*M)
    int calculateFreshOrangesAndAddRottenOranges(vector<vector<int>>& grid, vector<vector<int>>& vis, queue<tuple<int, int, int> >& q){

        int cnt=0;
        for(int i=0; i<grid.size(); i++){
            for(int j=0; j<grid[0].size(); j++){

                if(grid[i][j] == 2){
                    q.push({i, j, 0});
                    vis[i][j] = 2;
                }
                else if(grid[i][j] == 1) cnt++;
            }
        }

        return cnt;
    }

    int orangesRotting(vector<vector<int>>& grid){

        int N = grid.size();
        int M = grid[0].size();

        vector<vector<int>> vis(N, vector<int>(M, 0));
        queue<tuple<int, int, int>> q;
        int freshOranges = calculateFreshOrangesAndAddRottenOranges(grid, vis, q);

        if(freshOranges == 0) return 0;

        int delta[4][2] = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};

        int time=0;
        while(!q.empty()){

            auto [row, col, tm] = q.front();
            q.pop();
            time = max(time, tm);

            for(int i=0; i<4; i++){
                int newRow = row + delta[i][0];
                int newCol = col + delta[i][1];

                if(newRow >= 0 && newRow < N && newCol >= 0 && newCol < M && grid[newRow][newCol] == 1 && vis[newRow][newCol] != 2){
                    q.push({newRow, newCol, tm + 1});
                    vis[newRow][newCol] = 2;
                    freshOranges--;
                }
            }
        }

        if(freshOranges == 0) return time;
        return -1;
    }
};

File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    private int calculateFreshOrangesAndAddRottenOranges(int[][] grid, int[][] vis, Queue<int[]> q){
        
        int cnt=0;
        for(int i=0; i<grid.length; i++){
            for(int j=0; j<grid[0].length; j++){
                
                if(grid[i][j] == 2){
                    q.add(new int[]{i, j});
                    vis[i][j] = 2;
                }
                else{
                    vis[i][j] = 0;
                }
                
                if(grid[i][j] == 1) cnt++;
            }
        }
        
        return cnt;
    }
    
    public int orangesRotting(int[][] grid) {
        
        int N = grid.length;
        int M = grid[0].length;
        
        int[][] vis = new int[N][M];
        Queue<int[]> q = new LinkedList<>();
        int freshOranges = calculateFreshOrangesAndAddRottenOranges(grid, vis, q);
        
        if(freshOranges == 0) return 0;
        int[][] delta = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
        
        int time=0;
        while(q.size()>0 && freshOranges > 0){
            
            int size = q.size();
            while(size-- > 0){
                
                int row = q.peek()[0];
                int col = q.peek()[1];
                q.remove();
            
                for(int i=0; i<delta.length; i++){
                int nrow = row + delta[i][0];
                int ncol = col + delta[i][1];
                
                    if(nrow >= 0 && ncol >= 0 && nrow < N && ncol < M && grid[nrow][ncol] == 1 && vis[nrow][ncol] != 2){
                        q.add(new int[]{nrow, ncol});
                        vis[nrow][ncol] = 2;
                        freshOranges--;
                    }
                }
            }
            time++;
        }
        
        if(freshOranges == 0) return time;
        return -1;
    }
}